tensorboard/loader.py

```diff
@@ -0,0 +1,414 @@
+             not self._is_tapped and
+             not self._records):
+        self._no_soup_for_you.wait()
+      return self._get_record()
+
+  def close(self):
+    """Closes event log reader if open.
+
+    If the i/o thread is running, this method blocks until it has been
+    shut down.
+
+    Further reads are not permitted after this method is called.
+
+    Raises:
+      Exception: To propagate any exceptions that may have been thrown
+          by the close operation in the other thread. If an exception
+          is thrown, then all subsequent calls to this method will
+          rethrow that same exception.
+    """
+    with self._lock:
+      if not self._is_closed:
+        self._is_closed = True
+        if not self._is_started:
+          return
+        self._please_sir_i_want_some_more.notify()
+      while self._reader is not None:
+        self._no_soup_for_you.wait()
+      if self._close_exception is not None:
+        raise self._close_exception
+
+  def _get_record(self):
+    if self._read_exception is not None:
+      raise self._read_exception
+    if not self._records:
+      return None
+    record = self._records.popleft()
+    self._buffered -= len(record.record)
+    if self._should_wakeup():
+      self._please_sir_i_want_some_more.notify()
+    return record
+
+  def _run(self):
+    while True:
+      with self._lock:
+        while not self._should_wakeup():
+          self._please_sir_i_want_some_more.wait()
+        if self._is_closed:
+          try:
+            self._reader.close()
+            tf.logging.debug('Closed')
+          except Exception as e:
+            self._close_exception = e
+            tf.logging.debug('Close failed: %s', e)
+          self._reader = None
+          self._no_soup_for_you.notify_all()
+          return
+        if not self._should_rebuffer():
+          continue
+        # Calculate a good amount of data to read outside the lock.
+        # The less we have buffered, the less re-buffering we'll do.
+        # We want to minimize wait time in the other thread.
+        x = float(self._buffered)
+        c = float(self._read_ahead)
+        s = BufferedRecordReader.READ_AHEAD_AGGRESSION
+        want = int(min(c - x, s/c * x**s + 1))
+      tf.logging.debug('Waking up to read %s bytes', util.add_commas(want))
+      records = []
+      read_exception = None
+      try:
+        while want > 0:
+          record = self._reader.get_next_record()
+          if record is None:
+            break
+          records.append(record)
+          want -= len(record.record)
+        if self._should_stat():
+          self._stat()
+      except Exception as e:
+        read_exception = e
+      with self._lock:
+        if read_exception is not None:
+          self._read_exception = read_exception
+          self._is_closed = True
+        elif not records:
+          self._is_tapped = True
+        else:
+          for record in records:
+            self._records.append(record)
```

Comment:
It's to avoid repeatedly appending to the end of a `deque` (which would be accidentally quadratic if this were an `ArrayDeque`, but is not technically quadratic because CPython's `deque` implementation uses a blocked-linked-list). (Analogously, `n = int(1e6); x = range(n); for i in range(n, 2 * n): x.append(i)` vs `x = range(n); x.extend(range(n, 2 * n))`: the former is not quadratic because of amortized resizing, but certainly does have the potential to be noticeably slower due to intermediate resizes.)

Doing it as you have written is totally fine with me.